### PR TITLE
Add GCM client

### PR DESCRIPTION
**Pull request by wangmengyan95**
Add GCM client for push notification. For the `expirationTime`, we use a fixed length string in go, do we need to follow this?

### CODE REVIEW

**Comment by wangmengyan95** [comment]
cc @fantasist for review

File: GCM.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ * @returns {Object} A promise which is resolved after we get results from gcm
+ */
+var generateGCMPayload = function(coreData, pushId, timeStamp, expirationTime) {
+  var payloadData =  {
+    'time': timeStamp.toString(),
+    'push_id': pushId,
+    'data': JSON.stringify(coreData)
+  }
+  var payload = {
+    priority: 'high',
```
Comment (grantland) [inline GCM.js]: We shouldn't default to high on GCM

Comment (wangmengyan95) [inline GCM.js]: Any reason besides the battery concern?

Comment (grantland) [inline GCM.js]: It's also not the current default for hosted Parse

**Comment by bnham** [comment]
Mostly LGTM.

For the date, you get to define the parameters here right? It seems like expiration_date should just be a regular JS date or null.

For priority, here was our plan for supporting GCM priorities in the API server:

 1. If the user specifies `"priority": "high"` to /1/push, then send with high priority.
 2. If the user specifices `"priority": "normal"` to /1/push, then send with normal priority.
 3. Otherwise, send with high priority only if the `alert` key is present in the payload.

If you do anything with priority here, I vote for either always sending with low priority (which we currently do in the API), or using the simple policy described in (3).

**Comment by wangmengyan95** [comment]
@bnham , for the `expiration_time`, we handle the `expiration_time` convertion in `push.js`, so `expiration_time` is a valid date in Unix epoch time in milliseconds here

File: GCM.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ * Generate the gcm payload from the data we get from api request.
+ * @param {Object} coreData The data field under api request body
+ * @param {String} pushId A random string
+ * @param {Number} timeStamp A number whose format is the Unix Epoch
+ * @param {Number|undefined} expirationTime A number whose format is the Unix Epoch or undefined
+ * @returns {Object} A promise which is resolved after we get results from gcm
+ */
+var generateGCMPayload = function(coreData, pushId, timeStamp, expirationTime) {
+  var payloadData =  {
+    'time': timeStamp.toString(),
```
Comment (bnham) [inline GCM.js]: This should be an ISO8601 date, not an integer.

Comment (wangmengyan95) [inline GCM.js]: :+1:

**Pull request merged**